pkg/scheduler/redis_scheduler.go

```diff
@@ -54,39 +60,103 @@ type RedisScheduler struct {
 // other cozy-stack processes to schedule jobs.
 func NewRedisScheduler(client *redis.Client) *RedisScheduler {
 	return &RedisScheduler{
-		client:  client,
-		stopped: make(chan struct{}),
+		client: client,
 	}
 }
 
 func redisKey(infos *TriggerInfos) string {
 	return infos.Domain + "/" + infos.TID
 }
 
+func eventsKey(domain string) string {
+	return "events-" + domain
+}
+
 // Start a goroutine that will fetch triggers in redis to schedule their jobs
 func (s *RedisScheduler) Start(b jobs.Broker) error {
 	s.broker = b
+	s.stopped = make(chan struct{})
+	go s.eventDispatcher()
+	go s.pollLoop()
+	return nil
+}
+
+func (s *RedisScheduler) pollLoop() {
+	ticker := time.NewTicker(pollInterval)
+	for {
+		select {
+		case <-s.stopped:
+			ticker.Stop()
+			return
+		case <-ticker.C:
+			now := time.Now().UTC().Unix()
+			if err := s.Poll(now); err != nil {
+				log.Warnf("[scheduler] Failed to poll redis: %s", err)
+			}
+		}
+	}
+}
+
+func (s *RedisScheduler) eventDispatcher() {
+	eventsCh := make(chan *realtime.Event, 100)
 	go func() {
-		ticker := time.NewTicker(pollInterval)
+		c := realtime.GetHub().SubscribeAll()
 		for {
 			select {
 			case <-s.stopped:
-				ticker.Stop()
+				c.Close()
+				close(eventsCh)
 				return
-			case <-ticker.C:
-				now := time.Now().UTC().Unix()
-				if err := s.Poll(now); err != nil {
-					log.Warnf("[Scheduler] Failed to poll redis: %s", err)
-				}
+			case event := <-c.Read():
+				eventsCh <- event
 			}
 		}
 	}()
-	return nil
+	for i := 0; i < eventLoopSize; i++ {
+		go s.eventLoop(eventsCh)
+	}
+}
+
+func (s *RedisScheduler) eventLoop(ch <-chan *realtime.Event) {
+	for event := range ch {
+		key := eventsKey(event.Domain)
+		m, err := s.client.HGetAll(key).Result()
+		if err != nil {
+			log.Errorf("[scheduler] Could not fetch redis set %s: %s",
+				key, err.Error())
+			continue
+		}
+		for triggerID, args := range m {
+			rule, err := permissions.UnmarshalRuleString(args)
+			if err != nil {
+				log.Warnf("[scheduler] Coud not unmarshal rule %s: %s",
+					key, err.Error())
+				continue
+			}
+			if !eventMatchPermission(event, &rule) {
+				continue
+			}
+			t, err := s.Get(event.Domain, triggerID)
+			if err != nil {
+				log.Warnf("[scheduler] Could not fetch @event trigger %s %s: %s",
+					event.Domain, triggerID, err.Error())
+				continue
+			}
+			_, _, err = s.broker.PushJob(t.(*EventTrigger).Trigger(event))
```

Comment:
There will be some minor conflicts with my PR, but it should be easy to fix them